test/e2e/storage/api/testsuite.go

```diff
@@ -0,0 +1,129 @@
+	testName := fmt.Sprintf("[Testpattern: %s]%s %s%s", pattern.Name, pattern.FeatureTag, tsInfo.Name, tsInfo.FeatureTag)
+	ginkgo.Context(testName, func() {
+		ginkgo.BeforeEach(func() {
+			// skip all the invalid combination of driver and pattern
+			SkipInvalidDriverPatternCombination(driver, pattern)
+			// skip the unsupported test pattern and driver combination specific for this TestSuite
+			suite.SkipUnsupportedTests(driver, pattern)
+		})
+		// actually define the tests
+		// at this step the testsuite should not worry about if the pattern and driver
+		// does not fit for the whole testsuite. But driver&pattern check
+		// might still needed for specific independent test cases.
+		suite.DefineTests(driver, pattern)
+	})
+}
+
+// DefineTestSuites defines tests for all testpatterns and all testSuites for a driver
+func DefineTestSuites(driver TestDriver, tsInits []func() TestSuite) {
+	for _, testSuiteInit := range tsInits {
+		suite := testSuiteInit()
+		for _, pattern := range suite.GetTestSuiteInfo().TestPatterns {
+			RegisterTests(suite, driver, pattern)
+		}
+	}
+}
+
+// TestSuiteInfo represents a set of parameters for TestSuite
+type TestSuiteInfo struct {
+	Name               string              // name of the TestSuite
+	FeatureTag         string              // featureTag for the TestSuite
+	TestPatterns       []TestPattern       // Slice of TestPattern for the TestSuite
+	SupportedSizeRange e2evolume.SizeRange // Size range supported by the test suite
+}
+
+// SkipInvalidDriverPatternCombination will skip tests if the combination of driver, and testpattern
+// is not compatible to be tested. This function will be called in the RegisterTests() to make
+// sure all the testsuites we defined are valid.
+//
+// Whether it needs to be skipped is checked by following steps:
+// 0. Check with driver SkipUnsupportedTest
+// 1. Check if volType is supported by driver from its interface
+// 2. Check if fsType is supported
+//
+// Test suites can also skip tests inside their own skipUnsupportedTests function or in
+// individual tests.
+func SkipInvalidDriverPatternCombination(driver TestDriver, pattern TestPattern) {
+	dInfo := driver.GetDriverInfo()
+	var isSupported bool
+
+	// 0. Check with driver specific logic
+	driver.SkipUnsupportedTest(pattern)
+
+	// 1. Check if Whether volType is supported by driver from its interface
+	switch pattern.VolType {
+	case InlineVolume:
+		_, isSupported = driver.(InlineVolumeTestDriver)
+	case PreprovisionedPV:
+		_, isSupported = driver.(PreprovisionedPVTestDriver)
+	case DynamicPV, GenericEphemeralVolume:
+		_, isSupported = driver.(DynamicPVTestDriver)
+	case CSIInlineVolume:
+		_, isSupported = driver.(EphemeralTestDriver)
+	default:
+		isSupported = false
+	}
+
+	if !isSupported {
+		e2eskipper.Skipf("Driver %s doesn't support %v -- skipping", dInfo.Name, pattern.VolType)
+	}
+
+	// 2. Check if fsType is supported
+	if !dInfo.SupportedFsType.Has(pattern.FsType) {
```

Comment:
Replying to @jingxu97 question on https://github.com/kubernetes/kubernetes/pull/96538#discussion_r523150057

When we define an external driver, and e2e framework reads from the config file, it will always add "" so that makes sure it support default fsType
https://github.com/kubernetes/kubernetes/blob/da75c266487c52425d3eb07df30229aea04bb28e/test/e2e/storage/external/external.go#L185